### PR TITLE
Internvl fix

### PR DESCRIPTION
**Pull request by remi-or**
This PR brings to main a fix that was added in `amd-hf-ci-branch` here: https://github.com/huggingface/transformers/pull/38540
The fix aims to correct a failure of `torch.compile` when compiling a torchvision `resize` function with an uint8 input. The failing tests were: 
```
tests/models/internvl/test_video_processor_internvl.py::test_can_compile_fast_video_processor
tests/models/qwen2_vl/test_video_processing_qwen2_vl.py::test_can_compile_fast_video_processor
```
This PR also add some Expectations for AMD machines.

### CODE REVIEW

File: src/transformers/image_processing_utils_fast.py [hunk 1]
```diff
@@ -279,8 +280,34 @@ def resize(
                 "Size must contain 'height' and 'width' keys, or 'max_height' and 'max_width', or 'shortest_edge' key. Got"
                 f" {size}."
             )
+        # This is a workaround to avoid a bug in torch.compile when dealing with uint8 on AMD MI3XX GPUs
+        # Tracked in PyTorch issue: https://github.com/pytorch/pytorch/issues/155209
+        # TODO: remove this once the bug is fixed (detected with torch==2.7.0+git1fee196, torchvision==0.22.0+9eb57cd)
+        if torch.compiler.is_compiling() and is_rocm_platform():
+            return self.compile_friendly_resize(image, new_size, interpolation, antialias)
```
Comment (ydshieh) [inline src/transformers/image_processing_utils_fast.py:288]: could you share what tests would need this?

Why we have compiling in image processing stage ..?

File: src/transformers/models/bridgetower/image_processing_bridgetower_fast.py [hunk 2]
```diff
@@ -165,13 +165,18 @@ def resize(
             raise ValueError(f"The `size` dictionary must contain the key `shortest_edge`. Got {size.keys()}")
         shorter = size.shortest_edge
         longer = int(1333 / 800 * shorter)
-        output_size = get_resize_output_image_size(
+        output_height, output_width = get_resize_output_image_size(
```
Comment (ydshieh) [inline src/transformers/models/bridgetower/image_processing_bridgetower_fast.py:168]: could you explain why such changes here and below?

File: tests/models/internvl/test_modeling_internvl.py [hunk 3]
```diff
@@ -705,6 +705,7 @@ def test_llama_small_model_integration_forward(self):
                 ("xpu", 3): torch.tensor([-9.8750, -0.5703, 1.4297, -10.3125, -10.3125], dtype=torch.float16),
                 ("cuda", 7): torch.tensor([-9.8750,  -0.4861,   1.4648, -10.3359, -10.3359], dtype=torch.float16),
                 ("cuda", 8): torch.tensor([-9.8906,  -0.4995,   1.4473, -10.3359, -10.3438], dtype=torch.float16),
+                ("rocm", (9, 5)): torch.tensor([ -9.8906,  -0.4976,   1.4502, -10.3359, -10.3438], dtype=torch.float16),
```
Comment (ydshieh) [inline tests/models/internvl/test_modeling_internvl.py:708]: ok

**Comment by ydshieh** [comment]
It's better to explain a bit about the changes. (maybe it's explained in the PR for `amd-hf-ci-branch`).

Also since this is not only test files, I feel it's better for other team members (who work a lot on image stuffs ) to take a look too.

File: src/transformers/image_processing_utils_fast.py (hunk 1 above)
Comment (remi-or) [reply src/transformers/image_processing_utils_fast.py:288]: Added the failing tests to the PR description! As for why there is compiling, I am not sure -- probably helps speedup the processing stage

File: src/transformers/models/bridgetower/image_processing_bridgetower_fast.py (hunk 2 above)
Comment (remi-or) [reply src/transformers/models/bridgetower/image_processing_bridgetower_fast.py:168]: The idea here is to use the parent's class `resize` function rather than `torchvision` equivalent directly. To do that, we unpack the `output_size` to then use the `output_height` and `output_width` to build a SizeDict which is needed for `super().resize`

**Comment by remi-or** [comment]
Yes you are right, my bad! I added more context to the PR description. And as for other team members, I agree, tagging @yonigozlan and @zucchini-nlp as they already reviewed the last PR.

File: src/transformers/models/qwen2_vl/video_processing_qwen2_vl.py [hunk 4]
```diff
@@ -244,8 +244,10 @@ def _preprocess(
                     min_pixels=min_pixels,
                     max_pixels=max_pixels,
                 )
-                stacked_videos = F.resize(
-                    stacked_videos, size=(resized_height, resized_width), interpolation=interpolation
+                stacked_videos = self.resize(
+                    image=stacked_videos,
+                    size=SizeDict(height=resized_height, width=resized_width),
+                    interpolation=interpolation,
```
Comment (zucchini-nlp) [inline src/transformers/models/qwen2_vl/video_processing_qwen2_vl.py:256]: Qwen2-vl image processor as well 😄

**Review by zucchini-nlp — COMMENTED** [review]
Since it was discussed already in side branch, lgtm! 

Tiny nit: I see many other models calling `F.resize` instead of `super()` which aren't fixed in this PR (e.g. pixtral, convnext). Do we need to fix them as well, or are the tests not failing for them?

File: src/transformers/models/qwen2_vl/video_processing_qwen2_vl.py (hunk 4 above)
Comment (remi-or) [reply src/transformers/models/qwen2_vl/video_processing_qwen2_vl.py:256]: Made the change in `image_processing_qwen2_vl_fast.py` but not in `image_processing_qwen2_vl.py` because it has a `input_data_format` argument I am not sure about -- did you mean to change it in both?

Comment (zucchini-nlp) [reply src/transformers/models/qwen2_vl/video_processing_qwen2_vl.py:256]: No, only the fast processor. Slow processors don't rely on `torchvision` :)

**Comment by remi-or** [comment]
We only looked at tests that failed on Mi3XX that passed on the T4 CI, so either (1) the tests did not pass on the T4 (2) they were skipped on MI3XX or (3) they passed for some weird reason (no compilation, etc). I can look into those tests next

**Comment by zucchini-nlp** [comment]
Yeah, let's check those tests as well. In case the compile issue is not affecting all models equally under same settings, I think we should not need the workaround

File: src/transformers/image_processing_utils_fast.py [hunk 5]
```diff
@@ -279,8 +280,34 @@ def resize(
...
+    ) -> "torch.Tensor":
+        """
+        A wrapper around `F.resize` so that it is compatible with torch.compile when the image is a uint8 tensor.
+        """
+        if image.dtype == torch.uint8:
+            image = image.float() / 256
+            image = F.resize(image, new_size, interpolation=interpolation, antialias=antialias)
+            image = image * 256
+            image = image.masked_fill(image > 255, 255)
+            image = image.masked_fill(image < 0, 0)
```
Comment (ivarflakstad) [inline src/transformers/image_processing_utils_fast.py]: Did boolean masking get properly fixed in pytorch or does it still break the graph unless you set the `capture_dynamic_output_shape_ops = True` flag? ([ref](https://github.com/pytorch/vision/pull/8171))
In any case using `where` is supposedly more efficient.

**Review by yonigozlan — APPROVED** [review]
LGTM as well! Agreed with @zucchini-nlp, if we can address the issue of using F.resize instead of self.resize (or super().resize) for all fast image processors that would be great, might be a bit out of scope for this PR though

**Comment by ydshieh** [comment]
run-slow: bridgetower, llava_next, llava_onevision, qwen2_vl, qwen2_vl

**Comment by ydshieh** [comment]
let's wait the above CI runs

**Comment by remi-or** [comment]
> Yeah, let's check those tests as well. In case the compile issue is not affecting all models equally under same settings, I think we should not need the workaround

Tested with pixtral, and the test `test_can_compile_fast_image_processor` passes when `F.resize` output size is smaller than the input size (for instance, here input size is `[1, 3, 224, 244]` and output size is `[24, 24]`) but when we pick an output size larger (tried w/ `[448, 448]`) we get the same failures. So it seems only the up-sample code path is broken, which is why some tests have an issue and some dont. Will update the issue with that information

**Comment by ydshieh** [comment]
run-slow: bridgetower, llava_next, llava_onevision, qwen2_vl, qwen2_vl

**Comment by zucchini-nlp** [comment]
Great @remi-or , thanks for digging!

**Comment by ydshieh** [comment]
slow tests look good. I think we could merge?

**Comment by zucchini-nlp** [comment]
Yep, I also think we need to replace `F.resize` with out resize in other models, even though the tests are passing. In actual inference with rocm we aren't guaranteed to have resized shape smaller than the original shape

**Comment by remi-or** [comment]
Should we take care of it in this PR or should we leave that for a PR with a larger scope?

File: src/transformers/image_processing_utils_fast.py (hunk 5 above)
Comment (remi-or) [reply src/transformers/image_processing_utils_fast.py]: Did not know about this, changed.

**Review by ivarflakstad — APPROVED** [review]
